Improve MSVC C++ modules compatibility for nlohmann_json

banderzhm opened the pull request:
- Keep complete module interface with all exports
- Export all standard types: json, basic_json, adl_serializer, json_pointer, ordered_json, ordered_map
- Export detail namespace symbols for advanced usage
- Ensure compatibility with MSVC C++20/23 modules
- Maintain full functionality without breaking changes

banderzhm committed:
Improve MSVC C++ modules compatibility for nlohmann_json

- Keep complete module interface with all exports
- Export all standard types: json, basic_json, adl_serializer, json_pointer, ordered_json, ordered_map
- Export detail namespace symbols for advanced usage
- Ensure compatibility with MSVC C++20/23 modules
- Maintain full functionality without breaking changes

Signed-off-by: yudaichen <250074249@qq.com>

nlohmann commented:
Clang-Tidy issues are unrelated. I'll fix them.

banderzhm commented:
This is the most compatible approach for the MVSC platform for now. Please consider a code merge.

nlohmann commented:
As I wrote: there is currently an issue that makes the CI fail. I can only merge this once the CI is green.

nlohmann commented:
Please rebase to have a green CI. Please also check https://github.com/nlohmann/json/pull/4952/checks?check_run_id=52780702512

nlohmann commented:
Thanks! Please also check [DCO](https://github.com/nlohmann/json/pull/4952/checks?check_run_id=53390764478)

banderzhm commented:
Thank you for your patience. I'm new to this contribution workflow.

nlohmann commented on src/modules/json.cppm:
```diff
@@ -12,3 +12,21 @@ export namespace nlohmann {
+export using NLOHMANN_JSON_NAMESPACE::ordered_map;
+export using NLOHMANN_JSON_NAMESPACE::to_string;
+
+NLOHMANN_JSON_NAMESPACE_END
```
There is a newline missing at the end of the file.

nlohmann reviewed (changes requested):
To verify the module is actually working with MSVC, the test `ci_module_cpp20` needs to be added to the Windows workflow. Please edit `.github/workflows/windows.yml` and add this test:

```yaml
  ci_module_cpp20:
    runs-on: windows-latest
    steps:
      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
      - name: Run CMake (Debug)
        run: cmake -S . -B build -G "Visual Studio 17 2022" -DJSON_CI=ON -DCMAKE_CXX_FLAGS="/permissive- /std:c++latest /utf-8 /W4 /WX"
      - name: Build
        run: cmake --build build --config Debug --target ci_module_cpp20
```

This was taken from #4768.

banderzhm committed:
Add C++20 module Windows CI test and fix file ending

- Add ci_module_cpp20 job to Windows workflow to verify module support with MSVC

- Add missing newline at end of json.cppm file

Signed-off-by: yudaichen <250074249@qq.com>

nlohmann commented on .github/workflows/windows.yml:
```diff
@@ -112,3 +112,12 @@ jobs:
+      - name: Run CMake (Debug)
+        run: cmake -S . -B build -G "Visual Studio 17 2022" -DJSON_CI=ON -DCMAKE_CXX_FLAGS="/permissive- /std:c++latest /utf-8 /W4 /WX"
+      - name: Build
+        run: cmake --build build --config Debug --target ci_module_cpp20
```
Nitpick: please add a newline to the end of the file.

banderzhm commented:
OKOK . I have already added it

nlohmann reviewed (approved):
Looks good to me.

nlohmann commented:
Thanks a lot!
